### PR TITLE
Add `std::lower_bound` support

### PR DESCRIPTION
**Pull request by unterumarmung**

### CODE REVIEW

File: include/sortcheck.h [hunk 1]
```diff
@@ -199,6 +199,58 @@ inline bool binary_search_checked_full(_ForwardIterator __first,
   return binary_search_checked_full(__first, __last, __val, compare, file, line);
 }
 
+template<typename _ForwardIterator, typename _Tp, typename _Compare>
+inline _ForwardIterator lower_bound_checked(_ForwardIterator __first,
+                                  _ForwardIterator __last,
+                                  const _Tp &__val, _Compare __comp,
+                                  const char *file, int line) {
```
Comment (yugr) [inline include/sortcheck.h]: Could you align params here and in other functions?

File: include/sortcheck.h [hunk 2]
```diff
@@ -199,6 +199,58 @@ inline bool binary_search_checked_full(_ForwardIterator __first,
...
+  // Ordered
+  if (__first != __last) {
+    bool is_prev_less = true;
+    unsigned pos = 0;
+    for (_ForwardIterator it = __first; it != __last; ++it, ++pos) {
+      bool is_less = *it < __val;
```
Comment (yugr) [inline include/sortcheck.h]: `__comp(*it, __val)`? Would you mind fixing this for `binary_search_checked` as well?

File: src/SortChecker.cpp [hunk 3]
```diff
@@ -114,19 +114,32 @@ class Visitor : public RecursiveASTVisitor<Visitor> {
...
     // https://en.cppreference.com/w/cpp/named_req/Compare
     CMP_FUNC_NUM
-  } CompareFunction;
+  };
+
+  bool isKindOfBinarySearch(CompareFunction func) {
+    switch (func) {
+    case CMP_FUNC_BINARY_SEARCH:
+      LLVM_FALLTHROUGH;
+    case CMP_FUNC_LOWER_BOUND:
```
Comment (yugr) [inline src/SortChecker.cpp:137]: Why not just
```
    case CMP_FUNC_BINARY_SEARCH:
    case CMP_FUNC_LOWER_BOUND:
```
? Fallthrough annotations are normally omitted in such cases.

File: src/SortChecker.cpp [hunk 4]
```diff
@@ -44,7 +44,7 @@ class Visitor : public RecursiveASTVisitor<Visitor> {
     return E;
   }
 
-  void replaceCallee(DeclRefExpr *DRE, const char *Replacement) const {
+  void replaceCallee(DeclRefExpr *DRE, llvm::StringRef Replacement) const {
     SourceRange Range = {DRE->getBeginLoc(), DRE->getEndLoc()};
     RW.ReplaceText(Range, Replacement);
```
Comment (yugr) [inline src/SortChecker.cpp:56]: I was originally worried that `Replacement` will be cached inside the rewriter but it looks like it makes a copy so it's ok.

**Review by yugr — COMMENTED** [review]
Cool stuff! I've left minor nits.

File: src/SortChecker.cpp (hunk 3 above)
Comment (unterumarmung) [reply src/SortChecker.cpp:137]: Well, didn't fallthrough annotations added for such cases? 
For example, clang source code:
https://github.com/llvm/llvm-project/blob/be7e938e2712f3d84d1775ecdb1bc6784ea114f7/clang/lib/Interpreter/IncrementalParser.cpp#L68-L73

File: include/sortcheck.h (hunk 2 above)
Comment (unterumarmung) [reply include/sortcheck.h]: Yes, it is a copy paste error
It seems, it is already fixed for `binary_search_checked`

File: src/SortChecker.cpp (hunk 3 above)
Comment (yugr) [reply src/SortChecker.cpp:137]: There instances of another style too: https://github.com/llvm/llvm-project/blob/be7e938e2712f3d84d1775ecdb1bc6784ea114f7/clang/lib/CodeGen/CGException.cpp#L154 and AFAIK it's much more common, both in LLVM and beyond...

File: include/sortcheck.h [hunk 5]
```diff
@@ -199,6 +199,58 @@ inline bool binary_search_checked_full(_ForwardIterator __first,
...
+      if (is_less && !is_prev_less) {
+        std::ostringstream os;
+        os << file << ':' << line << ": unsorted range "
+           << "in position " << pos;
+        report_error(os.str());
+      }
+      is_prev_less = is_less;
+    }
+  }
```
Comment (yugr) [inline include/sortcheck.h]: We should move this code to `check_sorted` to avoid code dup with `binary_search`.

File: src/SortChecker.cpp (hunk 3 above)
Comment (unterumarmung) [reply src/SortChecker.cpp:137]: Ok, now I get what you mean
Will fix

File: include/sortcheck.h (hunk 5 above)
Comment (unterumarmung) [reply include/sortcheck.h]: I say we should deal with it in a separate PR

**Comment by unterumarmung** [comment]
@yugr I addressed your review comments in a separate commit for an easy review process 
Feel free to squash it if you decide to merge it

**Comment by unterumarmung** [comment]
Some of errors that CI has is 
```
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "builtin.cpp"
No compilation database found in /home/runner/work/sortcheckxx/sortcheckxx/tests/builtin or any parent directory
fixed-compilation-database: Error while opening fixed database: No such file or directory
json-compilation-database: Error while opening JSON database: No such file or directory
Running without flags.
```
I encountered it too
Do you know how to fix it?

**Comment by yugr** [comment]
> I encountered it too Do you know how to fix it?

I'd love to fix it but it's in fact harmless. Your build crashed due to missing perms:
```
tests/runtests.sh: 14: tests/lower_bound/run.sh: Permission denied
```

File: include/sortcheck.h (hunk 5 above)
Comment (yugr) [reply include/sortcheck.h]: Do you have plans to implement this refactor?

**Comment by unterumarmung** [comment]
> > I encountered it too Do you know how to fix it?
> 
> I'd love to fix it but it's in fact harmless. Your build crashed due to missing perms:
> 
> ```
> tests/runtests.sh: 14: tests/lower_bound/run.sh: Permission denied
> ```

Well, yes, but I don't what's is wrong
I commited the `run.sh` with
```
chmod +x run.sh
```
And I've verified it

**Comment by yugr** [comment]
> > I encountered it too Do you know how to fix it?
> 
> I'd love to fix it but it's in fact harmless.

Fixed.

**Comment by yugr** [comment]
> And I've verified it

Cloning you repository I see
```
-rw-rw-r-- 1 yugr yugr 865 янв 29 23:44 run.sh
```
so there must be something about your local system settings...

**Comment by unterumarmung** [comment]
Hi, @yugr!
I fixed the CI problem

File: include/sortcheck.h (hunk 5 above)
Comment (yugr) [reply include/sortcheck.h]: I've factored out the `check_ordered` code and fixed some bugs in it. Could you use it here instead of code dup?

Comment (unterumarmung) [reply include/sortcheck.h]: Done!

**Review by yugr — APPROVED** [review]

**Comment by yugr** [comment]
Tanks!

**Pull request merged**